File: curriculum/07_rest_api_auth/node-impl/src/__tests__/app.test.ts
import { describe, expect, it } from 'vitest';
import jwt from 'jsonwebtoken';
import { buildApp, defaultConfig } from '../app';

const strong = 'CorrectHorse1Battery';

interface PublicUser { id: string; email: string; display_name: string; roles: string[]; status: string }
interface Success<T> { data: T; request_id: string }
interface Failure { error: { code: string; message: string; details: Array<{ field: string; reason: string }> }; request_id: string }
interface LoginData { access_token: string; refresh_token: string; token_type: string; expires_in_seconds: number; refresh_expires_in_seconds: number; user: PublicUser }
interface RefreshData { access_token: string; refresh_token: string }
const bodyOf = <T>(response: { json: () => unknown }): T => response.json() as T;
const testApp = () => buildApp({ config: { passwordIterations: 100 } });

async function registered() {
  const built = testApp();
  const register = await built.app.inject({ method: 'POST', url: '/v1/auth/register', payload: { email: 'Ada@Example.com', password: strong, display_name: 'Ada' } });
  const user = bodyOf<Success<{ user: PublicUser }>>(register).data.user;
  const login = await built.app.inject({ method: 'POST', url: '/v1/auth/login', payload: { email: 'ada@example.com', password: strong } });
  return { ...built, user, tokens: bodyOf<Success<LoginData>>(login).data };
}

describe('REST API auth Node implementation', () => {
  it('registers normalized users without exposing password material and audits it', async () => {
    const { app, store } = testApp();
    const res = await app.inject({ method: 'POST', url: '/v1/auth/register', headers: { 'x-request-id': 'req_test' }, payload: { email: 'Learner@Example.com', password: strong, display_name: 'Learner' } });
    expect(res.statusCode).toBe(201);
    const body = bodyOf<Success<{ user: PublicUser }>>(res);
    expect(body).toMatchObject({ data: { user: { email: 'learner@example.com', roles: ['user'], status: 'active' } }, request_id: 'req_test' });
    expect(JSON.stringify(body)).not.toContain(strong);
    expect([...store.users.values()][0].passwordHash).not.toBe(strong);
    expect(store.audits.map((entry) => entry.action)).toContain('user_registered');
  });

  it('rejects invalid and duplicate registration with structured errors', async () => {
    const { app } = testApp();
    const invalid = await app.inject({ method: 'POST', url: '/v1/auth/register', payload: { email: 'bad', password: 'weak', extra: true } });
    expect(invalid.statusCode).toBe(400);
    expect(bodyOf<Failure>(invalid).error.code).toBe('VALIDATION_FAILED');
    await app.inject({ method: 'POST', url: '/v1/auth/register', payload: { email: 'a@example.com', password: strong } });
    const duplicate = await app.inject({ method: 'POST', url: '/v1/auth/register', payload: { email: 'A@example.com', password: strong } });
    expect(duplicate.statusCode).toBe(409);
    expect(bodyOf<Failure>(duplicate).error.code).toBe('EMAIL_ALREADY_REGISTERED');
  });

  it('logs in, issues verifiable JWT claims, and hides credential failure cause', async () => {
    const { app, store } = await registered();
    const ok = await app.inject({ method: 'POST', url: '/v1/auth/login', payload: { email: 'ada@example.com', password: strong } });
    expect(ok.statusCode).toBe(200);
    const body = bodyOf<Success<LoginData>>(ok).data;
    const decoded = jwt.verify(body.access_token, defaultConfig.jwtSecret, { issuer: defaultConfig.issuer, audience: defaultConfig.audience, clockTimestamp: Date.parse('2026-06-17T00:00:00.000Z') / 1000 }) as jwt.JwtPayload;
    expect(decoded.sub).toBeDefined();
    expect(decoded.roles).toEqual(['user']);
    expect(decoded.jti).toContain('jti_');
    const bad = await app.inject({ method: 'POST', url: '/v1/auth/login', payload: { email: 'none@example.com', password: 'WrongPassword1' } });
    expect(bad.statusCode).toBe(401);
    expect(bodyOf<Failure>(bad).error.code).toBe('INVALID_CREDENTIALS');
    expect(store.audits.map((entry) => entry.action)).toContain('login_failed');
  });

  it('enforces authentication, admin RBAC, and ownership update policy', async () => {
    const { app, store, user, tokens } = await registered();
    const missing = await app.inject({ method: 'GET', url: '/v1/users' });
    expect(missing.statusCode).toBe(401);
    const malformed = await app.inject({ method: 'GET', url: '/v1/users', headers: { authorization: 'Bearer not-a-jwt' } });
    expect(malformed.statusCode).toBe(401);
    const denied = await app.inject({ method: 'GET', url: '/v1/users', headers: { authorization: `Bearer ${tokens.access_token}` } });
    expect(denied.statusCode).toBe(403);
    const self = await app.inject({ method: 'PUT', url: `/v1/users/${user.id}`, headers: { authorization: `Bearer ${tokens.access_token}` }, payload: { display_name: 'Ada L.' } });
    expect(self.statusCode).toBe(200);
    const forbidden = await app.inject({ method: 'PUT', url: `/v1/users/${user.id}`, headers: { authorization: `Bearer ${tokens.access_token}` }, payload: { roles: ['admin'] } });
    expect(forbidden.statusCode).toBe(403);
    const stored = store.users.get(user.id);
    if (!stored) throw new Error('missing user');
    stored.roles = ['admin'];
    const adminLogin = await app.inject({ method: 'POST', url: '/v1/auth/login', payload: { email: 'ada@example.com', password: strong } });
    const adminTokens = bodyOf<Success<LoginData>>(adminLogin).data;
    const listed = await app.inject({ method: 'GET', url: '/v1/users?limit=10', headers: { authorization: `Bearer ${adminTokens.access_token}` } });
    expect(listed.statusCode).toBe(200);
    expect(bodyOf<Success<{ users: PublicUser[] }>>(listed).data.users).toHaveLength(1);
    const invalidLimit = await app.inject({ method: 'GET', url: '/v1/users?limit=101', headers: { authorization: `Bearer ${adminTokens.access_token}` } });
    expect(invalidLimit.statusCode).toBe(400);
  });

  it('rotates refresh tokens and detects replay', async () => {
    const { app, store, tokens } = await registered();
    const refreshed = await app.inject({ method: 'POST', url: '/v1/auth/refresh', payload: { refresh_token: tokens.refresh_token } });
    expect(refreshed.statusCode).toBe(200);
    expect(bodyOf<Success<RefreshData>>(refreshed).data.refresh_token).not.toBe(tokens.refresh_token);
    const replay = await app.inject({ method: 'POST', url: '/v1/auth/refresh', payload: { refresh_token: tokens.refresh_token } });
    expect(replay.statusCode).toBe(401);
    expect(bodyOf<Failure>(replay).error.code).toBe('REFRESH_TOKEN_REPLAYED');
    expect(store.audits.map((entry) => entry.action)).toContain('refresh_replayed');
  });

  it('serves health and structured unsupported version errors', async () => {
    const { app } = testApp();
    const health = await app.inject({ method: 'GET', url: '/healthz' });
    expect(bodyOf<Success<{ status: string }>>(health).data.status).toBe('ok');
    const unsupported = await app.inject({ method: 'GET', url: '/v2/users' });
    expect(unsupported.statusCode).toBe(404);
    expect(bodyOf<Failure>(unsupported).error.code).toBe('UNSUPPORTED_API_VERSION');
  });
});

File: curriculum/07_rest_api_auth/go-impl/internal/authapi/app_test.go
package authapi

import (
	"bytes"
	"encoding/json"
	"io"
	"log/slog"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

const strongPassword = "CorrectHorse1Battery"

func testApp() *App {
	return NewApp(DefaultConfig(), FixedClock{Value: time.Date(2026, 6, 17, 0, 0, 0, 0, time.UTC)}, slog.New(slog.NewTextHandler(io.Discard, nil)))
}
func request(app *App, method, path string, body interface{}, token string) *httptest.ResponseRecorder {
	var reader io.Reader
	if body != nil {
		raw, _ := json.Marshal(body)
		reader = bytes.NewReader(raw)
	}
	req := httptest.NewRequest(method, path, reader)
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	if token != "" {
		req.Header.Set("Authorization", "Bearer "+token)
	}
	rr := httptest.NewRecorder()
	app.Router().ServeHTTP(rr, req)
	return rr
}
func registerAndLogin(t *testing.T) (*App, map[string]interface{}, map[string]interface{}) {
	t.Helper()
	app := testApp()
	reg := request(app, http.MethodPost, "/v1/auth/register", map[string]interface{}{"email": "Ada@Example.com", "password": strongPassword, "display_name": "Ada"}, "")
	if reg.Code != http.StatusCreated {
		t.Fatalf("register failed: %d %s", reg.Code, reg.Body.String())
	}
	login := request(app, http.MethodPost, "/v1/auth/login", map[string]interface{}{"email": "ada@example.com", "password": strongPassword}, "")
	if login.Code != http.StatusOK {
		t.Fatalf("login failed: %d %s", login.Code, login.Body.String())
	}
	var regBody, loginBody map[string]interface{}
	_ = json.Unmarshal(reg.Body.Bytes(), &regBody)
	_ = json.Unmarshal(login.Body.Bytes(), &loginBody)
	return app, regBody, loginBody
}

func TestRegisterValidationHashingAndAudit(t *testing.T) {
	app := testApp()
	res := request(app, http.MethodPost, "/v1/auth/register", map[string]interface{}{"email": "Learner@Example.com", "password": strongPassword, "display_name": "Learner"}, "")
	if res.Code != http.StatusCreated {
		t.Fatalf("got %d", res.Code)
	}
	if strings.Contains(res.Body.String(), strongPassword) {
		t.Fatal("response leaked password")
	}
	user := app.Store().FindUserByEmail("learner@example.com")
	if user == nil || user.PasswordHash == strongPassword {
		t.Fatal("password was not hashed")
	}
	if len(app.Store().Audits) != 1 || app.Store().Audits[0].Action != "user_registered" {
		t.Fatal("missing registration audit")
	}

	invalid := request(app, http.MethodPost, "/v1/auth/register", map[string]interface{}{"email": "bad", "password": "weak", "extra": true}, "")
	if invalid.Code != http.StatusBadRequest {
		t.Fatalf("expected validation failure")
	}
	dup := request(app, http.MethodPost, "/v1/auth/register", map[string]interface{}{"email": "LEARNER@example.com", "password": strongPassword}, "")
	if dup.Code != http.StatusConflict {
		t.Fatalf("expected duplicate conflict")
	}
}

func TestLoginJwtAndCredentialFailure(t *testing.T) {
	app, _, loginBody := registerAndLogin(t)
	data := loginBody["data"].(map[string]interface{})
	token := data["access_token"].(string)
	parsed, err := jwt.Parse(token, func(_ *jwt.Token) (interface{}, error) { return []byte(DefaultConfig().JWTSecret), nil }, jwt.WithIssuer(DefaultConfig().Issuer), jwt.WithAudience(DefaultConfig().Audience), jwt.WithTimeFunc(func() time.Time { return time.Date(2026, 6, 17, 0, 0, 0, 0, time.UTC) }))
	if err != nil || !parsed.Valid {
		t.Fatalf("jwt invalid: %v", err)
	}
	claims := parsed.Claims.(jwt.MapClaims)
	if claims["sub"] == "" || claims["jti"] == "" {
		t.Fatal("missing claims")
	}
	bad := request(app, http.MethodPost, "/v1/auth/login", map[string]interface{}{"email": "none@example.com", "password": "WrongPassword1"}, "")
	if bad.Code != http.StatusUnauthorized || !strings.Contains(bad.Body.String(), "INVALID_CREDENTIALS") {
		t.Fatalf("bad credential response: %d %s", bad.Code, bad.Body.String())
	}
}

func TestRBACAndOwnership(t *testing.T) {
	app, reg, login := registerAndLogin(t)
	user := reg["data"].(map[string]interface{})["user"].(map[string]interface{})
	userID := user["id"].(string)
	token := login["data"].(map[string]interface{})["access_token"].(string)
	if request(app, http.MethodGet, "/v1/users", nil, "").Code != http.StatusUnauthorized {
		t.Fatal("missing token should be 401")
	}
	if request(app, http.MethodGet, "/v1/users", nil, token).Code != http.StatusForbidden {
		t.Fatal("user list should require admin")
	}
	if request(app, http.MethodPut, "/v1/users/"+userID, map[string]interface{}{"display_name": "Ada L."}, token).Code != http.StatusOK {
		t.Fatal("self display update should pass")
	}
	if request(app, http.MethodPut, "/v1/users/"+userID, map[string]interface{}{"roles": []string{"admin"}}, token).Code != http.StatusForbidden {
		t.Fatal("self role update should fail")
	}
	stored := app.Store().GetUser(userID)
	stored.Roles = []Role{RoleAdmin}
	app.Store().SaveUser(stored)
	adminLogin := request(app, http.MethodPost, "/v1/auth/login", map[string]interface{}{"email": "ada@example.com", "password": strongPassword}, "")
	var body map[string]interface{}
	_ = json.Unmarshal(adminLogin.Body.Bytes(), &body)
	adminToken := body["data"].(map[string]interface{})["access_token"].(string)
	if request(app, http.MethodGet, "/v1/users?limit=10", nil, adminToken).Code != http.StatusOK {
		t.Fatal("admin list should pass")
	}
}

func TestRefreshRotationReplayHealthAndVersion(t *testing.T) {
	app, _, login := registerAndLogin(t)
	refresh := login["data"].(map[string]interface{})["refresh_token"].(string)
	first := request(app, http.MethodPost, "/v1/auth/refresh", map[string]interface{}{"refresh_token": refresh}, "")
	if first.Code != http.StatusOK {
		t.Fatalf("refresh failed: %d", first.Code)
	}
	replay := request(app, http.MethodPost, "/v1/auth/refresh", map[string]interface{}{"refresh_token": refresh}, "")
	if replay.Code != http.StatusUnauthorized || !strings.Contains(replay.Body.String(), "REFRESH_TOKEN_REPLAYED") {
		t.Fatalf("replay should be detected")
	}
	if request(app, http.MethodGet, "/healthz", nil, "").Code != http.StatusOK {
		t.Fatal("health failed")
	}
	if request(app, http.MethodGet, "/v2/users", nil, "").Code != http.StatusNotFound {
		t.Fatal("unsupported version should be 404")
	}
}
